### PR TITLE
sdk/metric: Record measurements when context is done

### PR DESCRIPTION
**Pull request by pellared**
Fixes https://github.com/open-telemetry/opentelemetry-go/issues/4750

I think that the `ctx` passed to the "metric recording" API should be used only for "context propagation" meaning e.g. correlation with tracing. I do not think it should cancel the recordings.

Similarly we can still record spans when the context is canceled.

Some issues related with the current behavior:

- https://github.com/open-telemetry/opentelemetry-collector/pull/9048
- https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4634
- https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4356#discussion_r1350178596

### CODE REVIEW

**Comment by pellared** [comment]
This is also ready for review.

**Review by MrAlias — COMMENTED** [review]
If a user passes a context to `Add` or `Record` it would seem like a reasonable expectation that its deadline and cancellation would be honored. And their call to these methods would not hold much beyond the cancellation. This seems to be the intended behavior of a context and something ubiquitous throughout the Go communities use of them.

Why should our default behavior be different here?

It seems like a caller has the responsibility to manage the context and pass one in the state they desire. If they want to pass the values in a context that is already cancelled why isn't it their responsibility then to not copy the values to a new context?

**Comment by MrAlias** [comment]
> If a user passes a context to `Add` or `Record` it would seem like a reasonable expectation that its deadline and cancellation would be honored. And their call to these methods would not hold much beyond the cancellation. This seems to be the intended behavior of a context and something ubiquitous throughout the Go community's use of them.
> 
> Why should our default behavior be different here?
> 
> It seems like a caller has the responsibility to manage the context and pass one in the state they desire. If they want to pass the values in a context that is already cancelled why isn't it their responsibility then to copy the values to a new context?

It seems this was the intent of [`WithoutCancel`](https://pkg.go.dev/context#WithoutCancel). I think that should be prioritized instead of accommodating here.

**Comment by pellared** [comment]
> It seems this was the intent of [`WithoutCancel`](https://pkg.go.dev/context#WithoutCancel). I think that should be prioritized instead of accommodating here.

Nice idea. I have missed this new method when reading release notes👍 The only problem is that it is not available in Go 1.20 that we still support. Using it in go-contrib once we drop support for Go 1.20.

I will leave this proposal opened until the next SIG.

**Comment by MrAlias** [comment]
> > It seems this was the intent of [`WithoutCancel`](https://pkg.go.dev/context#WithoutCancel). I think that should be prioritized instead of accommodating here.
> 
> Nice idea. I have missed this new method when reading release notes👍 The only problem is that it is not available in Go 1.20 that we still support. Using it in go-contrib once we drop support for Go 1.20.
> 
> I will leave this proposal opened until the next SIG.

The implementation looks easy enough to copy if that solution is wanted before Go 1.20: https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/context/context.go;l=566-595

**Comment by pellared** [comment]
SIG meeting:
1. It will be safer to not change the behavior.
2. The user may expect the context cancelation to be handled.
3. The user may always use a context which ignores the cancelation.

**Comment by pellared** [comment]
For reference:

> Canceling the context should not affect record processing.
> (Among other things, log messages may be necessary to debug a
> cancellation-related problem.)

Source: https://pkg.go.dev/log/slog#Handler

**Comment by pellared** [comment]
More issues related with the current behavior: 
- https://github.com/open-telemetry/opentelemetry-collector/pull/9048
- https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4634

**Comment by pellared** [comment]
The PR was discussed during  SIG meeting on 2023-12-07.

We seemed to have an agreement that the change is acceptable.
However, we decided that the decision should be documented in `CONTRIBUTING.md` as ignoring a context cancelation is not something usual. 

We also double-checked that cancelation of the context passed to `span.Start` is ignored as well.

**Comment by jmacd** [comment]
Here is my position: https://github.com/open-telemetry/opentelemetry-collector/pull/9048#issuecomment-1843247202

**Comment by carrbs** [comment]
If we merge this, should we also remove the `withoutCancel` implementation and usage in otelgrpc?

https://github.com/open-telemetry/opentelemetry-go-contrib/blob/6009b3188da4f25eb8d2e65ffeefb3bf4b7c3415/instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go#L148

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ### Changed
 
+- Record synchronous measurements when the passed context is canceled instead of dropping in `go.opentelemetry.io/otel/sdk/metric`.
+  If you do not want to make measurment in such scenario then you need to handle it yourself.
+  You can do it e.g. by adding `if ctx.Err() != nil` condition before making the measurement.
+  However, we believe that you rather not want it as the telemetry may be necessary to diagnose a cancellation-related problem. (#4671)
```
Comment (dashpole) [inline CHANGELOG.md:26]: not sure this last sentence is necessary

**Review by dashpole — APPROVED** [review]

**Comment by carrbs** [comment]
> > If we merge this, should we also remove the `withoutCancel` implementation and usage in otelgrpc?
> 
> 
> 
> @carrbs Correct. We can prepare draft PR and make it ready for review after this one is merged.

I can work on that. Should there be a new issue opened for this in the contrib repo beforehand?

**Comment by pellared** [comment]
> > > If we merge this, should we also remove the `withoutCancel` implementation and usage in otelgrpc?
> > 
> > 
> > @carrbs Correct. We can prepare draft PR and make it ready for review after this one is merged.
> 
> I can work on that. Should there be a new issue opened for this in the contrib repo beforehand?

You can just create a PR and reference this PR in its description.

**Review by MadVikingGod — APPROVED** [review]

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -591,6 +591,20 @@ this.
 
 [^3]: https://github.com/open-telemetry/opentelemetry-go/issues/3548
 
+### Honoring context cancellation
```
Comment (MrAlias) [inline CONTRIBUTING.md:594]:
```suggestion
### Ignoring context cancellation
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -591,6 +591,20 @@ this.
...
+
+The implementation of OpenTelemetry API that is responsible for telemetry recording
+(starting a span, making a synchronous metric measurement, emitting a log)
+must not honor the cancelation of the passed context.
```
Comment (MrAlias) [inline CONTRIBUTING.md:598]:
```suggestion
OpenTelemetry API implementations need to ignore the cancellation of the context they are
passed when recording a value (e.g. starting a span, recording a measurement, emitting a log).
Recording methods should not return an error describing the cancellation state of the context
when they complete, nor should they abort any work.
```

**Comment by pellared** [comment]
@MrAlias Thanks for the suggestions. I made some (I think that 2) little corrections in your suggestions. Can you double-check?

**Review by MrAlias — APPROVED** [review]

**Pull request merged**